### ISSUE TITLE
NordVPN stopped working

### ISSUE THREAD

**Issue by Arathonk** [Zomboided/service.vpn.manager#393]
Since today NordVPN stopped working, I reinstalled and correctly filled in my login, but its giving me an message saying it can't connect.
Logfile: https://paste.kodi.tv/uqopeqagad

**Comment by Zomboided** [comment]
Thanks for the debug, I can see that it doesn't connect with an auth
error.  I have no idea why this is and I don't have a Nord subscription to
know if something has changed so I'm not sure I can add anything to this
right now.

**Comment by Arathonk** [comment]
Is there anything I can do to help you fix it?
By the way I just tried the User Defined option, and that way I actually can login for some reason..
Logfile: https://paste.kodi.tv/kunihemepi

**Comment by Zomboided** [comment]
You're failing to log into the Nord API service to be able to get the
'best' server based on your location.  This will return an ovpn file, which
is then used to log into the VPN.  You're "just" manually doing the last
piece...the issue is the with the Nord API piece.  This is your best
approach for now, but Nord change servers often so you'll be changing this
semi-regularly.

I'd need to know what Nord has changed and see their new API, I'm not sure
that's something you can help with.

**Comment by peno64** [comment]
I am also using NordVPN on kodi with this addon and I just tried it and I have no problem. I tried two times to connect and disconnect and that worked ok.
@Arathonk Did you change something to your NordVPN account? Change password, enable two factor auth or something like that?

**Comment by Arathonk** [comment]
No not at all, I checked and two factor auth is off and password has been the same for a long time, I can login on other devices no problem, so that's what makes it weird. 
Also because the user defined works, and that also uses the login info, password and two factor shouldn't be the issue.

**Comment by peno64** [comment]
From your log file I see you are have a simular configuration as me with quite the same addons. I also use LibreElec with kodi 20.1. I run this on a rpi4. I guess you are from The Netherlands, I am from Belgium. 
I will try to connect to a Dutch vpn.

**Comment by peno64** [comment]
Just tried it. Netherlands and Belgium works without a problem here.
No special characters in your password? Not to long or to short? Maybe something changed that these don't work anymore... On the other had you do say that it works with user defined. Have you tried editing settings.xml from the zomboided addon and look there if the password is really as you expect it?

**Comment by peno64** [comment]
Isn't this strange in the log:

2023-04-03 14:54:06.862 T:1069     info <general>: 2023-04-03 14:52:46 WARNING: file '/storage/.kodi/addons/service.vpn.manager/NordVPN/pass.txt' is group or others accessible
2023-04-03 14:54:06.862 T:1069     info <general>: 2023-04-03 14:52:46 OpenVPN 2.5.8 armv8a-libreelec-linux-gnueabihf [SSL (OpenSSL)] [LZO] [LZ4] [EPOLL] [AEAD] built on Mar 18 2023
2023-04-03 14:54:06.862 T:1069     info <general>: 2023-04-03 14:52:46 library versions: OpenSSL 3.0.8 7 Feb 2023, LZO 2.10
2023-04-03 14:54:06.863 T:1069     info <general>: 2023-04-03 14:52:46 Error reading username from Auth authfile: /storage/.kodi/addons/service.vpn.manager/NordVPN/pass.txt
2023-04-03 14:54:06.863 T:1069     info <general>: 2023-04-03 14:52:46 Exiting due to fatal error

And then it says:

2023-04-03 14:54:15.679 T:1210    error <general>: VPN Mgr : (alternativeNord.py) Couldn't authenticate with NordVPN
2023-04-03 14:54:15.679 T:1210    error <general>: VPN Mgr : (alternativeNord.py) API call was https://api.nordvpn.com/v1/users/tokens, username=wvannoort%40hotmail.com&password=********
2023-04-03 14:54:15.679 T:1210    error <general>: VPN Mgr : (alternativeNord.py) Response was 401 Unauthorized

It really looks like an issue with the password

**Comment by BrodjagaRatnik** [comment]
/storage/.kodi/addons/service.vpn.manager/NordVPN/pass.txt is empty after I connected to NordVPN

**Comment by kpisacic** [comment]
Something has really changed in NordVPN API. My installation also stopped working today. Credentials are correct, tested it from other VPN client and directly on nordaccount.com
I also tried through SoapUI - and response is:

```
HTTP/1.1 401 Unauthorized
Date: Mon, 03 Apr 2023 19:35:45 GMT
Transfer-Encoding: chunked
Connection: keep-alive
x-host-signature: ****
x-authorization: key-id="rsa-key-1",algorithm="rsa-sha256"
x-digest: *************
x-signature: *************
x-accept-before: 1680593745
CF-Cache-Status: DYNAMIC
Set-Cookie: *******; path=/; expires=Mon, 03-Apr-23 20:05:45 GMT; domain=.nordvpn.com; HttpOnly; Secure; SameSite=None
Strict-Transport-Security: max-age=31536000; includeSubDomains; preload
Server: cloudflare
CF-RAY: 7b23c1ac2dddc223-VIE
alt-svc: h3=":443"; ma=86400, h3-29=":443"; ma=86400

{ "errors": { "code": 100110, "message": "Invalid renew token" } }

```

It does not return "token" and "renew_token" as `authenticateNordVPN` expects.

Also tried with "service credentials" instead or real credentials - but behaves the same.

**Comment by peno64** [comment]
I just tried again and it still worked...
I rebooted my device and tried again and now it doesn't work anymore. So something was still in cache that made it still work and now after reboot it doesn't anymore...

**Comment by Zomboided** [comment]
Yeah, it caches the authentication token so you don't have to sign in each
time the API is used....guess this support is dead now, sorry all.

**Comment by peno64** [comment]
I am talking with a nordvpn person atm...

**Comment by peno64** [comment]
I wonder if this is the problem:
https://support.nordvpn.com/Connectivity/Linux/1905092252/How-to-log-in-to-NordVPN-on-Linux-with-a-token.htm

**Comment by Arathonk** [comment]
> I wonder if this is the problem: https://support.nordvpn.com/Connectivity/Linux/1905092252/How-to-log-in-to-NordVPN-on-Linux-with-a-token.htm

That would be worth trying right? Can that be edited into the pass.txt somehow?

**Comment by peno64** [comment]
I think the current tapi call
https://api.nordvpn.com/v1/users/tokens, username=xxxxx&password=********
must be replaced by something that provides the token.

Btw, the token  can be generated on the nordvpn page

**Comment by peno64** [comment]
Another strange thing is that on the page https://my.nordaccount.com/nl/dashboard/nordvpn/
it shows a user name and password
Both are however not the user name and password we used to use to login to nordvpn.
So I tried to use that information but it still doesn't work, unfortunately...

Anyone smarter than me who can figure this out? I have to go to sleep...

**Comment by tlaurion** [comment]
Broke for me too...
Login works, vpn on qubes vpn works as well
```
user@sys-nordvpn:~$ nordvpn connect
Connecting to Canada #1457 (ca1457.nordvpn.com)
You are connected to Canada #1457 (ca1457.nordvpn.com)!
```

**Comment by tlaurion** [comment]
As said previously, canadians and worldwide would profit from riseupvpn, but I can hardly let go on nordvpn as well.
Anything I can do to help here?

**Comment by mqttnodered** [comment]
Not sure what NordVPN have done but it also broke for me.
I did find a quick work around though. Reset all the OpenVPN settings (or just uninstall and reinstall) and use the UserDefined settings. 
Download a couple of .ovpn config files from NordVpn https://nordvpn.com/ovpn/ and also find your Service Credentials for manual setup. These are located on your dashboard page after you select the 'NordVPN' service. The Username and Password generated here will now be your login details for OpenVPN. 
You will need to import the .ovpn config files through the OpenVPN settings wizard for all the servers you have. Note that they will probably change over time so I'm not sure how stable this will be going forward. 

It looks like Nord allow your original Un and Pwd to log into any of the official NordVPN apps, but any third-party VPN client may have to use the alternative service credentials.

**Comment by Arathonk** [comment]
> Not sure what NordVPN have done but it also broke for me. I did find a quick work around though. Reset all the OpenVPN settings (or just uninstall and reinstall) and use the UserDefined settings. Download a couple of .ovpn config files from NordVpn https://nordvpn.com/ovpn/ and also find your Service Credentials for manual setup. These are located on your dashboard page after you select the 'NordVPN' service. The Username and Password generated here will now be your login details for OpenVPN. You will need to import the .ovpn config files through the OpenVPN settings wizard for all the servers you have. Note that they will probably change over time so I'm not sure how stable this will be going forward.
> 
> It looks like Nord allow your original Un and Pwd to log into any of the official NordVPN apps, but any third-party VPN client may have to use the alternative service credentials.

The original username and password worked for me when trying user defined tho

**Comment by Zomboided** [comment]
To fix this, I'd need access to the Nord API documentation, which I don't
have anymore.  Currently it uses your username and password to log on to
get a token to do 'things', but maybe that support link above is really
saying that the only way to get a token now is to generate it using their
website.  As @mqttnodered and others have said, the way to use Nord is now
is by downloading the ovpn files and using the wizard.  It's a shame
because the integration with their API made this one of the better options
to use.

**Comment by peno64** [comment]
I suggest that everyone of us sends a mail to support@nordvpn.com to report the issue to them to give some pressure.
Also say that their scripts are still python2 and don't work at all anymore with the current kodi versions.

**Comment by Arathonk** [comment]
> I suggest that everyone of us sends a mail to [support@nordvpn.com](mailto:support@nordvpn.com) to report the issue to them to give some pressure. Also say that their scripts are still python2 and don't work at all anymore with the current kodi versions.

Would you care giving us a template to mail them? As I have no idea how to compress this into an email.

**Comment by peno64** [comment]
Everyone should say it is his own words I think or it will look as spam

**Comment by kpisacic** [comment]
Would it be possible to utilize generated static token based authorization ( https://support.nordvpn.com/Connectivity/Linux/1905092252/How-to-log-in-to-NordVPN-on-Linux-with-a-token.htm ) ? 
To skip all username/password authorization to get temporary token, and instead use token manually generated on nordVpn page as described in instructions from some configuration variable?
Would it work to only return this token from some configuration in `def getTokenNordVPN()` ?

**Comment by Arathonk** [comment]
I would say it would be possible to implement that, but I don't have any coding experience, so I guess that's up to @Zomboided or anyone else that knows how to do this.

**Comment by MichaelLikesJazz** [comment]
I've send a message to NordVPN directly. However, I do not think that they'll change their process.

I am open to support here due to the fact that I am in the need of solution.

In addition, is there a way for a quick fix for LibreElec / Kodi etc. to use NordVPN until the fix of Zomboided?

**Comment by tlaurion** [comment]
Sent an email pointing to https://github.com/Zomboided/service.vpn.manager/issues/393#issuecomment-1495423130 with subject: "Upstream API change broken downstream VPN applications", urging them to revert the change done 2 days ago which broke access to raspberries in remote locations for me.

**Comment by mikfrase** [comment]
I started having same issue 2 days ago, RasPi4 with LibreELEC, Kodi 19.5 Matrix. I have sent NorrdVPN a support concern about this.  Following, I will attempt to resolve with what I am learning here.

**Comment by MichaelLikesJazz** [comment]
> > I've send a message to NordVPN directly. However, I do not think that they'll change their process.
> > I am open to support here due to the fact that I am in the need of solution.
> > In addition, is there a way for a quick fix for LibreElec / Kodi etc. to use NordVPN until the fix of Zomboided?
> 
> The quick fix is using user defined with your usual login, that works for me.

Many thanks! Works well as a workaround.

**Comment by kpisacic** [comment]
I have added code below into alternativeNord.py:
* added "return True" into start of authenticateNordVPN
* added reading of token from config "vpn_password" and returning token in start of getTokenNordVPN

I stored token from NordVPN profile page into configuration "vpn_password" - and then wizard and connectivity works fine without authorization (it downloads countries, chooses best server, connect). 

```
# diff alternativeNord.py.old alternativeNord.py
48a49,53
>     # start - workaround NordVPN issues
>     return True
>     # end - workaround NordVPN issues
> 
> 
118a124,129
>     # start - workaround for NordVPN API issues
>     addon = xbmcaddon.Addon(getID())
>     token = addon.getSetting("vpn_password")
>     return token
>     # end - workaround for NordVPN issues
> 
```

I think best solution would be:
* add "vpn_token" as configuration of addon (screen, config file), 
* if "vpn_token" populated, return True from authenticateNordVPN
* if "vpn_token" populated, return value of "vpn_token" from getTokenNordVPN

I would like to add this, but even do know some python, i don't have a clue how to add config parameter in Kodi addon.

br,k.

**Comment by peno64** [comment]
@kpisacic  

If @Zomboided doesn't have time or interest to make these change I would gladly do them. 

I have modified and even created some simple addons in the past so I am confident I can do this.

**Comment by FatFreddysDog** [comment]
Ive got the same issue contacted NordVPN and they wont tell me what theyve changed or help with 3rd party apps , however they did say this 
![image](https://user-images.githubusercontent.com/129892459/229882328-9957832a-8b1b-43b3-98f3-e9d52fd57813.png)

**Comment by FatFreddysDog** [comment]
also if i use user defined what folder do i drop the openvpn files into ?

**Comment by peno64** [comment]
@FatFreddysDog 
That doesn't matter. Just a place that can be accessed by kodi. The vpn manager wizard will ask for the file (directory or file) and you choose the location where you put it. The wizard will then copy it to the location where it is actually needed and even make modifications in it.

**Comment by FatFreddysDog** [comment]
> @FatFreddysDog That doesn't matter. Just a place that can be accessed by kodi. The vpn manager wizard will ask for the file (directory or file) and you choose the location where you put it. The wizard will then copy it to the location where it is actually needed and even make modifications in it.

brill cheers <goes away to firk it up>

**Comment by peno64** [comment]
@Zomboided 

> I don't know the best way of adding a token though, whether it's to replace the user ID/password, or whether it's to create a separate field which would appear if Nord is selected.

Wouldn't the easiest thing not be to rename the current label "User name" to "User name/Token" and provide an info label that it will be used as User name if there is a password entered and as token if no password is provided? Loging in with a user name without a password isn't possible anyway. Most probably not a solution that will win a price but it would require the least of work.

**Comment by Zomboided** [comment]
Yeah kinda, but you've seen the bug reports....expect a lot of  'where do I
get the token from' questions from non Nord users.  I would not do it this
way, even as a quick hack, I'd fiddle wtih the xml until it worked.  I
think it also offers you a way of telling Nord users via a line of text
instead of the password to 'Get token from your Nord account' or something
more meaningful.  Not everyone is invested in this stuff...they just want
to steal TV.

**Comment by peno64** [comment]
@Zomboided 

ok will look into it

**Comment by peno64** [comment]
@kpisacic 

I can confirm that your changes work great. Everything seems to work again as before with it.

Working on an update of the addon.

**Comment by Arathonk** [comment]
> @kpisacic
> 
> I can confirm that your changes work great. Everything seems to work again as before with it.
> 
> Working on an update of the addon.

That would be great! I have no clue what to do 😅

**Comment by quickstang** [comment]
Glad I'm not alone. Same thing happened to me on my Pi4. I'll try the workaround until there is an update of the addon, because this addon is a lifesaver and so easy to use.

**Comment by peno64** [comment]
Ok, I just created a pull request to update the addon so that it works again with NordVPN.
The version will becomme 7.0.2 Note that it will only work on kodi Matrix (19.x) and Nexus (20.x), not on kodi 19 (Leia) or lower.

Thanks to @kpisacic to find out how to fix it.

@Zomboided can you please accept the pull request and make a release of it.

Unfortunately it's not just updating the addon and all works again. There is a little more to it because it broke because NordVPN changed sommething to their api. It looks like it is no longer possible to login with user/password or at least we don't know how to do that with their changes.
Fortunately it is possible to login via a token (which was previously determined via the login with user/password). So instead of providing your login credentials to the addon, now the token must be provided.

To get a token you must login via a browser to your NordVPN account and go to https://my.nordaccount.com/dashboard/nordvpn/

Scoll a bit down and there you will see a section Access Token and a button create new access token. Do that and there you will get the possibility to create a token for (I think) one month or a token that will never expire. If you create one for one month then you will have to repeat below procedure again after that time. I created a token that never expires.

Copy that token and keep it somewhere safe such that you can get it back because when this window is closed you will not be able to visualize it again via NordVPN. If you forget that token you will have to revoke the token first and then create a new one.

Now once this is done and you updated the service vpn manager addon go into the settings of the addon and do the following:

- Settings, VPN Configuration
- Note that for VPN Provider NordVPN you will see it no longer asks for a user name and password but for a token.
- If NordVPN was previously already configured, the settings are disabled and you must click on "Click here to update VPN settings"
- Press Yes on message box
- Enter the token you obtained via https://my.nordaccount.com/dashboard/nordvpn/
- Go to Settings, VPN Connections and setup at least your first connection and possible others (if you previously had nordvpn configured you have lost this and you must define the connections again. Fortunately its the only thing you loose)
- Press Ok on the settings to save.

Another way, if you know how to do this, is edit settings.xml under userdata/addon_data/service.vpn.manager with a text editor and add thte token like this:

    <setting id="vpn_token">the token from the nordvpn website comes here</setting>

Save this and restart kodi. The advantage of this way is that you can copy/paste the token and you don't loose your existing configurations.

All should be ok now.

**Comment by mikfrase** [comment]
Looking forward in applying & testing this update on Matrix 19.5

**Comment by quickstang** [comment]
> Not sure what NordVPN have done but it also broke for me. I did find a quick work around though. Reset all the OpenVPN settings (or just uninstall and reinstall) and use the UserDefined settings. Download a couple of .ovpn config files from NordVpn https://nordvpn.com/ovpn/ and also find your Service Credentials for manual setup. These are located on your dashboard page after you select the 'NordVPN' service. The Username and Password generated here will now be your login details for OpenVPN. You will need to import the .ovpn config files through the OpenVPN settings wizard for all the servers you have. Note that they will probably change over time so I'm not sure how stable this will be going forward.
> 
> It looks like Nord allow your original Un and Pwd to log into any of the official NordVPN apps, but any third-party VPN client may have to use the alternative service credentials.

I got this method to work. I didn't have to change my username and password. All I did was use userdefined, downloaded a handful of .ovpn files, used filezilla to copy them in a new directory and loaded them in the add-on. Checked VPN status and it shows I'm connected.

I didn't have to use any service credentials. At least not at the moment.

**Comment by peno64** [comment]
The changes are on https://github.com/peno64/service.vpn.manager so until the PR is accepted you could install the zip from there.

**Comment by FatFreddysDog** [comment]
Brilliant you are a legend. Thanks

**Comment by Zomboided** [comment]
@peno64 Can you test https://github.com/Zomboided/service.vpn.manager/releases/tag/7.0.2 and I'll push it out on the repository.  They're your changes, plus some tracing.

**Comment by BrodjagaRatnik** [comment]
https://github.com/Zomboided/service.vpn.manager/releases/tag/7.0.2 works Ty.

**Comment by kpisacic** [comment]
@Zomboided , @peno64 - works also for me, many thnx!

**Comment by Zomboided** [comment]
ok, fixed the changelog (in github, for the next release) and pushed the change....thanks @peno64

**Comment by peno64** [comment]
@Zomboided 
I installed your released version and all is ok. 
Maybe it's a  good idea to pin this issue such that users find it before they create a new defect and also to see the instructions I wrote how to obtain the token.

**Comment by Arathonk** [comment]
Freshly installed 7.0.2 but its not working for me, neither with the wizard or via settings, I get "Please enter a user name and password. NordVPN requires them for authentication."

**Comment by peno64** [comment]
Are you sure you entered the correct token? How did you enter it? via copy/paste or enter it manually? And its the token from the nordvpn webpage you entered?

**Comment by Arathonk** [comment]
> Are you sure you entered the correct token? How did you enter it? via copy/paste or enter it manually? And its the token from the nordvpn webpage you entered?

Yes I copy/pasted it from the Nordvpn dashboard.

**Comment by peno64** [comment]
Does the setup page looks a bit like this:

![image](https://user-images.githubusercontent.com/32430988/230202815-75294852-2fd5-42e2-aff0-1c88287aff1c.png)

And the token is quite long? here it's 64 characters long.

**Comment by Arathonk** [comment]
> Does the setup page looks a bit like this:
> 
> ![image](https://user-images.githubusercontent.com/32430988/230202815-75294852-2fd5-42e2-aff0-1c88287aff1c.png)
> 
> And the token is quite long? here it's 64 characters long.

Yes it does look like that.

**Comment by Zomboided** [comment]
@peno64 This happens as soon as you validate a connection.  I bet there's
some code in there to check to see whether that's a userID/password.
There's probably code like that in multiple places.
You probably avoided it by editting your settings.xml which is totally not
recommended as I abuse the hell out of that file and previously there were
problems with settings being wiped out if things got hairy.

**Comment by peno64** [comment]
@Zomboided 
I did test it by entering the token via the settings screen and it worked fine. But it could be that the settings file also still contained the user name and password.
I will try it with an empty settings file and see what happens.

**Comment by Zomboided** [comment]
You want to fix line 56 of connect.py to check to see if a token exists if
the provider is nord, otherwise check the password/userID as it does today

**Comment by Zomboided** [comment]
Actually, I coded around this occurance already.  In vpnproviders.py, add
the "NordVPN" string to 'providers_no_pass' array and it'll probably all
just work fine.
I've committed a change for this in Git, but you'll need to test.

**Comment by peno64** [comment]
@Zomboided 
Ok, I will test it

**Comment by peno64** [comment]
@Zomboided 
I removed the settings.xml file, restarted kodi and setup the connection and indeed I also got the message "Please enter a user name and password. NordVPN requires them for authentication."

Then I applied your change and it fixes the problem.  I tried both manual settings and wizard starting from an empty settings file. manual settings is fine.

However wizard does not work. At a given moment tit says: "Click ok to create a VPN connection to NordVPN for user name ."
That is from common.py line 1280.
Maybe also here a test should be added if vpn_povider is NordVPN that it should say something different but probably not show the token.
Going further then fails. It says "Could not authenticate with VPN provider. Please check user name and password and try again. The strange thing is that it did not ask for a user rname, password or token in case of wizard.
At the beginning of the wizard it does say that it would ask for this during connection but it didn't. Maybe this is because o f your last change?

**Comment by peno64** [comment]
@Zomboided 

Indeed. If I remove  NordVPN from providers_no_pass the wizard setup does ask for a user name now. Which is also not correct.

Still need to do somo work here...

**Comment by peno64** [comment]
@Zomboided 

Ok I see how I can fix this. Will make a new pull request for this.

**Comment by peno64** [comment]
@Zomboided 

I created a new pull request. Tried with a removed settings file to setup nordvpn both via settings and via the wizard and all is good with it. It now also asks for a token in the wizard and it doesn't task or show traces of user name or password anymore. At least in the happy flow...

I also tested user defined both via settings and wizard and this also still works.

Sorry for not having tested this more carefully before.

**Comment by Zomboided** [comment]
I'd forgotten there was a new user wizard!  Ok, I've pushed your changes to
the repo and there's a release on Github.  I've done zero testing because I
have no time to spend on this today so I guess we're doing live user
testing here!

**Comment by peno64** [comment]
@Arathonk 
Can you test it with the new release (7.0.3)? If possible both with wizard and manual configuration via user setting as I see you also tried it yesterday with both and both failed for you. They should now work.
Thanks

**Comment by Arathonk** [comment]
> @Arathonk Can you test it with the new release (7.0.3)? If possible both with wizard and manual configuration via user setting as I see you also tried it yesterday with both and both failed for you. They should now work. Thanks

Both work now! Only when you use the manual configuration, after you've put in your token, first press okay, and they try to make a first connection, otherwise you can't make a first connection.

**Comment by peno64** [comment]
Credits must go more to @kpisacic who found out that the token from the  NordVPN page can be easily replace the user/password auth and how to implement it.

**Comment by hub880** [comment]
Hi. Are there plans to implement the same fix for version 6.3.7 ( kodi leia 18.9 / python2.7)?

**Comment by jim-cat** [comment]
Thanks for fixing!

**Comment by JuliusCaeser2030** [comment]
Works perfect thanks. Added a token. Connected no problem. Very easy to add profiles for other country servers...
Also, I think it's stating the obvious but I had OpenVPN and the NordVPN app but couldn't see the add token option until I installed the repository you are talking about here: https://github.com/Zomboided/repository.zomboided.plugins/releases/tag/1.0.1
Then it works like charm.

**Comment by fxleblond** [comment]
Hello. I can't connect with NordVPN. I successfully generated the Token, I entered it in the field after selecting NORDVPN from the list. At no time am I asked for a password! And when I launch the first connection, I get an error: 
"Error reading username from Authfile : /storage/.kodi/addons/service.vpn.manager/NordVPN/pass.txt".

 I tried to fill this file, either with the password associated with the Token, or the password associated with my email, and I still get the same error. I wonder what is the exact procedure for setting up NordVPN?

**Comment by peno64** [comment]
There must be something wrong with the token. Try to generate a new one and make sure that you check that it doesn't expire.
You should only have to enter the token in the settings. If it asks for the password or gives that error then a wrong token was entered.

**Comment by fxleblond** [comment]
> There must be something wrong with the token. Try to generate a new one and make sure that you check that it doesn't expire. You should only have to enter the token in the settings. If it asks for the password or gives that error then a wrong token was entered.

Thank you for your answer. Indeed I tried to generate a new token, and it works, without entering any login or password. Perfect !

**Comment by CryptoriousBIG** [comment]
Hi folks - I had to reinstall LibreElec on my Raspberry Pi and subsequently had to reinstall OpenVPN as well.  In doing so, I am no longer able to log-in and set up my NordVPN account.  I know a previous patch had been released to force the NordVPN sign-in details to be a token instead of a username and password combination but don't see this anymore.  Is there a new update I should be installing?

**Comment by peno64** [comment]
Are you sure you are using the latest version of this addon? When you choose NordVPN provider, it should ask for the token. If not I think you are using a wrong version.

**Comment by peno64** [comment]
Do you have a backup of your previous installation? If so and you can retrieve settings.xml from this addon and restore it on your new installation it should work immediately (after a restart of kodi). If not, also try to delete the current settings.xml, restart kodi and try again.

**Comment by CryptoriousBIG** [comment]
> Are you sure you are using the latest version of this addon? When you choose NordVPN provider, it should ask for the token. If not I think you are using a wrong version.

I pushed the .zip for version 7.0.4 to my Raspberry Pi using Putty. I also uninstalled and tried reinstalling OpenVPN by linking the Troypoint repo which has the Zomboided data in it but noted it has an earlier version of OpenVPN. Both ask for a username and password when selecting NordVPN.

**Comment by CryptoriousBIG** [comment]
> Do you have a backup of your previous installation? If so and you can retrieve settings.xml from this addon and restore it on your new installation it should work immediately (after a restart of kodi). If not, also try to delete the current settings.xml, restart kodi and try again.

Unfortunately not as I did a full wipe of my memory card to fix some issues I was having. I'll try wiping settings and restart. Can you confirm if 7.0.4 the latest and greatest for OpenVPN?

**Comment by peno64** [comment]
yes 7.0.4 is the latest. can you also check in kodi that 7.0.4 is actually installed.

**Comment by CryptoriousBIG** [comment]
> yes 7.0.4 is the latest. can you also check in kodi that 7.0.4 is actually installed.

Ok, that's the one I'm trying to install.  It seems maybe I haven't pathed the file correctly though with Putty.  I've only done this a few times in the past so maybe I'm making an error.  Here's what I've done so far, following the installation instructions posted in the wiki for the Zomboided OpenVPN:

1.  Pushed the zip file to LibreElec using Putty by typing the following "wget -O service.vpn.manager-7.0.4.zip https://github.com/Zomboided/service.vpn.manager/releases/tag/7.0.4"
2.  Installed the .zip in LibreElec
3.  Attempted to setup OpenVPN using the installation wizard but keep seeing a username and password input request for NordVPN

Upon deleting OpenVPN and trying this process again, the entire system (LibreElec) restarts on me when I attempt to install the .zip and OpenVPN doesn't show up when LibreElec boots back up.  Any ideas?

Am I entering the wget line in Putty incorrectly?  Apologies if I'm obviously missing something as I'm not an expert in doing this sort of thing quite yet.

**Comment by peno64** [comment]
You are not downloading the zip file but a html file with that command. That is also why kodi crashes with it.
This is the correct command: wget https://github.com/Zomboided/service.vpn.man
ager/releases/download/7.0.4/service.vpn.manager-7.0.4.zip

**Comment by CryptoriousBIG** [comment]
> You are not downloading the zip file but a html file with that command. That is also why kodi crashes with it. This is the correct command: wget https://github.com/Zomboided/service.vpn.man ager/releases/download/7.0.4/service.vpn.manager-7.0.4.zip

Perfect, that did it!  Thanks so much for your help!

**Comment by bam72** [comment]
I still couldn't get this to work using the Nord profile in the addon with the nord token, but using the User Defined option with downloaded openvpn files from nord and a pass.txt file with my credentials worked first time. I now have 5 nord servers set up and working perfectly with the latest version of KODI. I downloaded a load of server files from nord to my pc then uploaded them all to a folder on my pi and used this folder to import them all into openvpn manager. Its strange as Nord used to work fine with this addon then just stopped working. 
One thing I did do, which may or may not have been needed was to make a .crt and .key file for each server from the data within the .ovpn file.
Oh and I am using the 7.0.4 version from here not the 7.0.3 from the repository.
